src/cli/seperate_args.rs

```diff
@@ -0,0 +1,298 @@
+                let n = (*quotes >> 1) as usize;
+                let is_even = *quotes & 1 == 0;
+
+                arg.reserve(n);
+                for _ in 0..n {
+                    arg.push('"');
+                }
+                *quotes = 0;
+                *in_quotes = is_even;
+            }
+        }
+
+        for (index, c) in (&command[index..]).char_indices() {
+            match c {
+                '"' if in_first_argument => {
+                    in_quotes = !in_quotes;
+                }
+                '"' => {
+                    let is_backslash_escaped = backslashes % 2 == 1;
+                    push_backslashes(arg, &mut backslashes, true);
+
+                    if is_backslash_escaped {
+                        arg.push(c);
+                    } else if !in_quotes {
+                        in_quotes = true;
+                    } else {
+                        consecutive_quotes += 1;
+                    }
+                }
+                '\\' => {
+                    backslashes += 1;
+                    push_quotes(arg, &mut consecutive_quotes, &mut in_quotes);
+                }
+                // This filters empty arguments which doesn't really conform to the spec
+                // but we don't need them.
+                ' ' | '\t' if !in_quotes && matches!(last_char, ' ' | '\t') => {}
+                ' ' | '\t' => {
+                    push_backslashes(arg, &mut backslashes, false);
+                    push_quotes(arg, &mut consecutive_quotes, &mut in_quotes);
+
+                    if in_quotes {
+                        arg.push(c);
+                    } else {
+                        self.index += index + 1;
+                        self.in_first_argument = false;
+
+                        let mut result = String::with_capacity(arg.len());
+                        result.clone_from(arg);
+                        return Some(result);
+                    }
+                }
+                c => {
+                    push_backslashes(arg, &mut backslashes, false);
+                    push_quotes(arg, &mut consecutive_quotes, &mut in_quotes);
+                    arg.push(c);
+                }
+            }
+            last_char = c;
+        }
+        push_backslashes(arg, &mut backslashes, false);
+        push_quotes(arg, &mut consecutive_quotes, &mut in_quotes);
+
+        self.index = command.len() + 1;
+        if arg.is_empty() {
+            None
+        } else {
+            let mut result = std::mem::take(arg);
+            result.shrink_to_fit();
+
+            Some(result)
+        }
+    }
+}
+
+/// An iterator that parses a command as unix command-line arguments and returns them
+/// as [`String`]s.
+///
+/// Ported from cmake's [`kwsysSystem__ParseUnixCommand`][1].
+///
+/// [1]: https://github.com/Kitware/CMake/blob/859241d2bbaae83f06c44bc21ab0dbd38725a3fc/Source/kwsys/System.c#L94
+pub struct SeperateUnixArgs<'a> {
+    command: &'a str,
```

Comment:
We have checked and none of [shlex](https://github.com/comex/rust-shlex), [shellwords](https://crates.io/crates/shellwords) and [shell-words](https://github.com/tmiasko/shell-words) satisfies our needs, at least for Unix? Or is our concern that they are not that popular (and hence not well tested)?